src/unix/fs.c

```diff
@@ -120,6 +120,95 @@ static ssize_t uv__fs_fdatasync(uv_fs_t* req) {
 }
 
 
+static ssize_t uv__fs_clamp_max_path_len(ssize_t len) {
+  if (len == -1) {
+  #if defined(PATH_MAX)
+    len = PATH_MAX;
+  #else
+    len = 4096;
+  #endif
+  }
+
+  return len;
+}
+
+/*
+ * Computes the required size in bytes for directory entries read from the
+ * directory identified by the handle "dir".
+ *
+ * This code does not trust values of NAME_MAX that are less than
+ * 255, since some systems (including at least HP-UX) incorrectly
+ * define it to be a smaller value.
+ *
+ * This code was copied from
+ * http://womble.decadent.org.uk/readdir_r-advisory.html
+ * and slightly adapted.
+ */
+static size_t uv__fs_direntry_size(const DIR* dir) {
+  long name_max;
+  size_t name_end;
+
+  name_max = -1;
+  name_end = -1;
+
+#if defined(HAVE_FPATHCONF) && defined(HAVE_DIRFD) \
+    && defined(_PC_NAME_MAX)
+  name_max = fpathconf(dirfd(dirp), _PC_NAME_MAX);
+  if (name_max == -1)
+#   if defined(NAME_MAX)
+      name_max = (NAME_MAX > 255) ? NAME_MAX : 255;
+#   else
+      return (size_t)(-1);
```

Comment:
I think the idea was that `-1` would indicate an error.  Maybe `0` would be better?